rust/src/builders/certificate_builder.rs

```diff
@@ -0,0 +1,122 @@
+    match &cert_enum.0 {
+        // stake key registrations do not require a witness
+        CertificateEnum::StakeRegistration(_cert) => Ok(false),
+        CertificateEnum::StakeDeregistration(cert) => match cert.stake_credential().to_scripthash() {
+            Some(script_hash) => {
+                match *expected_hash == script_hash {
+                    true => Ok(true),
+                    false => return Err(JsError::from_str(&format!("Deregistration certificate contains wrong script hash. Expected {}, got {}.\n{:#?}", expected_hash, script_hash, cert.to_json()))),
+                }
+            },
+            None => return Err(JsError::from_str(&format!("Deregistration certificate contains public key hash. Expected script.\n{:#?}", cert.to_json()))),
+        },
+        CertificateEnum::StakeDelegation(cert) => match cert.stake_credential().to_scripthash() {
+            Some(script_hash) => {
+                match *expected_hash == script_hash {
+                    true => Ok(true),
+                    false => return Err(JsError::from_str(&format!("Delegation certificate contains wrong script hash. Expected {}, got {}.\n{:#?}", expected_hash, script_hash, cert.to_json()))),
+                }
+            },
+            None => return Err(JsError::from_str(&format!("Delegation certificate contains public key hash. Expected script.\n{:#?}", cert.to_json()))),
+        },
+        // can't register pools with scripts
+        CertificateEnum::PoolRegistration(_cert) => Ok(false),
+        CertificateEnum::PoolRetirement(_cert) => Ok(false),
+        // genesis keys are not scripts
+        CertificateEnum::GenesisKeyDelegation(_cert) => Ok(false),
+        // no witness as there is no single core node or genesis key that posts the certificate
+        CertificateEnum::MoveInstantaneousRewardsCert(_cert) => Ok(false),
+    }
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct CertificateBuilderResult {
+    cert: Certificate,
+    aggregate_witness: InputAggregateWitnessData,
+}
+
+#[wasm_bindgen]
+#[derive(Clone)]
+pub struct SingleCertificateBuilder {
+    cert: Certificate,
+}
+
+#[wasm_bindgen]
+impl SingleCertificateBuilder {
+    pub fn new(cert: &Certificate) -> Self {
+        Self {
+            cert: cert.clone(),
+        }
+    }
+
+    pub fn no_script(&self) -> Result<CertificateBuilderResult, JsError> {
+        let mut vkey_set = HashSet::<Ed25519KeyHash>::new();
+        add_cert_vkeys(&self.cert, &mut vkey_set)?;
+        Ok(CertificateBuilderResult {
+            cert: self.cert.clone(),
+            aggregate_witness: InputAggregateWitnessData::Vkeys(vkey_set.clone()),
+        })
+    }
+
+    pub fn native_script(&self, native_script: &NativeScript) -> Result<CertificateBuilderResult, JsError> {
+        let expected_hash = native_script.hash(ScriptHashNamespace::NativeScript);
+        check_cert_script_hash(&self.cert, &expected_hash)?;
```

Comment:
Will we need the `Ok(bool)` return type in other spots?

Reply:
Ah, I forgot to used the variable. My bad